hack/generate/changelog/gen-changelog.go

```diff
@@ -0,0 +1,331 @@
+	}
+
+	if err := updateChangelog(config{
+		File:    changelogFile,
+		Title:   tag,
+		Entries: entries,
+	}); err != nil {
+		log.Fatalf("failed to update CHANGELOG: %v", err)
+	}
+
+	if err := createMigrationGuide(config{
+		File:    filepath.Join(migrationDir, fmt.Sprintf("%s.md", tag)),
+		Title:   tag,
+		Entries: entries,
+	}); err != nil {
+		log.Fatalf("failed to update migration guide: %v", err)
+	}
+}
+
+func loadEntries(fragmentsDir string) ([]entry, error) {
+	files, err := ioutil.ReadDir(fragmentsDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read fragments directory: %w", err)
+	}
+
+	var entries []entry
+	for _, fragFile := range files {
+		if fragFile.Name() == "00-template.yaml" {
+			continue
+		}
+		if fragFile.IsDir() {
+			log.Warnf("Skipping directory %q", fragFile.Name())
+			continue
+		}
+		if filepath.Ext(fragFile.Name()) != ".yaml" || fragFile.IsDir() {
+			log.Warnf("Skipping non-YAML file %q", fragFile.Name())
+			continue
+		}
+		path := filepath.Join(fragmentsDir, fragFile.Name())
+		f, err := os.Open(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open fragment file %q: %w", fragFile.Name(), err)
+		}
+
+		decoder := yaml.NewDecoder(f)
+		fragment := fragment{}
+		if err := decoder.Decode(&fragment); err != nil {
+			return nil, fmt.Errorf("failed to parse fragment file %q: %w", fragFile.Name(), err)
+		}
+
+		if err := fragment.validate(); err != nil {
+			return nil, fmt.Errorf("failed to validate fragment file %q: %w", fragFile.Name(), err)
+		}
+
+		prNum, err := getPullRequest(path)
+		if err != nil {
+			log.Warnf("failed to get PR for fragment file %q: %v", fragFile.Name(), err)
+		}
+
+		if prNum != 0 {
+			for i, e := range fragment.Entries {
+				if e.PullRequest == nil {
+					fragment.Entries[i].PullRequest = &prNum
+				}
+			}
+		}
+
+		entries = append(entries, fragment.Entries...)
+	}
+	return entries, nil
+}
+
+func updateChangelog(c config) error {
+	changelog := map[entryKind][]string{}
+	for _, e := range c.Entries {
+		changelog[e.Kind] = append(changelog[e.Kind], e.toChangelogString())
+	}
+
+	var bb bytes.Buffer
+	order := []entryKind{
+		addition,
+		change,
+		removal,
+		deprecation,
+		bugfix,
+	}
+	bb.WriteString(fmt.Sprintf("## %s\n\n", c.Title))
+	for _, k := range order {
+		if entries, ok := changelog[k]; ok {
+			bb.WriteString(k.toHeader() + "\n\n")
+			for _, e := range entries {
+				bb.WriteString(e + "\n")
+			}
+			bb.WriteString("\n")
+		}
+	}
+
+	existingFile, err := ioutil.ReadFile(c.File)
+	if err != nil {
+		return fmt.Errorf("could not read CHANGELOG: %v", err)
+	}
+	bb.Write(existingFile)
+
+	if err := ioutil.WriteFile(c.File, bb.Bytes(), 0644); err != nil {
+		return fmt.Errorf("could not write CHANGELOG file: %v", err)
+	}
+	return nil
+}
+
+func createMigrationGuide(c config) error {
+	var bb bytes.Buffer
+
+	bb.WriteString("---\n")
+	bb.WriteString(fmt.Sprintf("title: %s\n", c.Title))
+
+	// TODO: sort these according to semver?
```

Comment:
Seems like a good plan. This could be tricky though. The weights in hugo are sorted smallest to largest, and I assume that we would want to order most recent first. 

The tricky part would come from the fact that the version components are all integers. The good news is that the weights can be huge numbers. 

Heres a first stab:
``` weight = 1000000 - (x * 10000) -  (y * 100) - z```
The idea is that x is 2 orders of magnitude bigger than y, which is 2 orders bigger than z, so this would work even if we release more than 10 of a y or z.

Reply:
This works up until y or z are > 99. Hopefully we never release 100 patch versions :grimacing:.

Reply:
Do we think we'll ever have pre-release CHANGELOGs and/or version migration guides?

If so, this becomes more complicated. If not, I'll add a check that will fail the generator unless the tag is a valid release semver.

Reply:
I ended up allowing up to `999.999.999`, and enforcing that the tag is valid semver, meaning no pre-release versions are allowed. We can revisit this if we ever decide to make a pre-release version.